File: app/src/main/java/com/levnovikov/postbus/root/home/prebooking/poi_selector/PoiSelectorView.java
package com.levnovikov.postbus.root.home.prebooking.poi_selector;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.example.core_geo.Point;
import com.levnovikov.postbus.R;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

/**
 * Created by lev.novikov
 * Date: 20/12/17.
 */

public class PoiSelectorView extends LinearLayout implements PoiSelectorInteractor.Presenter {

    public PoiSelectorView(Context context) {
        super(context);
    }

    public PoiSelectorView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public PoiSelectorView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Inject
    PoiSelectorInteractor interactor;

    private BehaviorSubject<Point> selectedPoiStream = BehaviorSubject.create();
    private BehaviorSubject<String> placesStream = BehaviorSubject.create();
    private Adapter adapter;

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        adapter = new Adapter((LayoutInflater) this.getContext().getSystemService( Context.LAYOUT_INFLATER_SERVICE ));
        initView();
        interactor.onGetActive();
    }

    private void initView() {
        final EditText editText = findViewById(R.id.poi_text);
        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {}

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {}

            @Override
            public void afterTextChanged(Editable editable) {
                placesStream.onNext(editable.toString());
            }
        });
        final RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    }

    @Override
    public Observable<Point> getSelectedPoi() {
        return selectedPoiStream;
    }

    @Override
    public Observable<String> getPlaceTitleStream() {
        return placesStream;
    }

    @Override
    public void updateSuggestions(List<Point> poiList) {
        adapter.setData(poiList);
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private TextView textView;

        ViewHolder(View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.title);
        }

        public void bind(Point point) {
            textView.setOnClickListener(v -> selectedPoiStream.onNext(point));
            textView.setText(point.title);
        }
    }

    class Adapter extends RecyclerView.Adapter<ViewHolder> {

        private LayoutInflater inflater;
        private List<Point> points = new ArrayList<>();

        Adapter(LayoutInflater inflater) {
            this.inflater = inflater;
        }

        void setData(List<Point> points) {
            this.points = points;
            notifyDataSetChanged();
        }

        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new ViewHolder(inflater.inflate(R.layout.poi_selector_item, parent, false));
        }

        @Override
        public void onBindViewHolder(ViewHolder holder, int position) {
            holder.bind(points.get(position));
        }

        @Override
        public int getItemCount() {
            return points.size();
        }
    }
}

File: app/src/main/java/com/levnovikov/postbus/root/home/prebooking/poi_selector/PoiSelectorView.kt
package com.levnovikov.postbus.root.home.prebooking.poi_selector

import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import com.example.core_geo.Point
import com.levnovikov.postbus.R
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import java.util.*
import javax.inject.Inject

/**
 * Created by lev.novikov
 * Date: 20/12/17.
 */

class PoiSelectorView : LinearLayout, PoiSelectorInteractor.Presenter {

    @Inject
    lateinit var interactor: PoiSelectorInteractor

    private val selectedPoiStream = BehaviorSubject.create<Point>()
    private val placesStream = BehaviorSubject.create<String>()
    private lateinit var adapter: Adapter

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        adapter = Adapter(this.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater)
        initView()
        interactor.onGetActive()
    }

    private fun initView() {
        val editText = findViewById<EditText>(R.id.poi_text)
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

            override fun afterTextChanged(editable: Editable) {
                placesStream.onNext(editable.toString())
            }
        })
        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(context)
    }

    override fun selectedPoi(): Observable<Point> {
        return selectedPoiStream
    }

    override fun placeTitleStream(): Observable<String> {
        return placesStream
    }

    override fun updateSuggestions(poiList: List<Point>) {
        adapter.setData(poiList)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val textView: TextView = itemView.findViewById(R.id.title)

        fun bind(point: Point) {
            textView.setOnClickListener { v -> selectedPoiStream.onNext(point) }
            textView.text = point.title
        }
    }

    inner class Adapter(private val inflater: LayoutInflater) : RecyclerView.Adapter<ViewHolder>() {
        private var points: List<Point> = ArrayList()

        fun setData(points: List<Point>) {
            this.points = points
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(inflater.inflate(R.layout.poi_selector_item, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.bind(points[position])
        }

        override fun getItemCount(): Int {
            return points.size
        }
    }
}
